### ISSUE TITLE
Read nifti files with matlab

### ISSUE THREAD

**Issue by Aghaeifar** [rordenlab/dcm2niix#210]
There is command 'niftiread' introduced in Matlab2017b to read nifti files.
When I convert DICOMs to nifti with the old tool dcm2nii, Matlab can read the generated nifti file, but when I do with dcm2niix, Matlab returns empty array. 
Are the process of nifti file creation different in these two tools?

Thanks.
Ali

**Comment by neurolabusc** [comment]
Both dcm2niix and my old tools generate standard NIfTI images that can be handled with popular tools. A nice way to see the contents of the NIfTI header is to use `fslhd`.  My newer software does populate some of the fields a bit differently (e.g. different strings stored for image description), but this should not cause issues.

I was not able to replicate any problems reading dcm2niix files using Matlab 2018a. Below is a test dataset from the [dcm_qa](https://github.com/neurolabusc/dcm_qa/tree/master/Ref) repository. I also tried ni.gz files as well using both `niftiread` and `niftiinfo`.  The NIfTI header is very compact and explicit (just 348 bytes), so it should be pretty easy to troubleshoot (from Matlab type `edit images.internal.nifti.niftiImage`). If you are using Matlab 2017b you may want to first try upgrading to 2018a. Perhaps it is a bug in Matlab that Mathworks has already fixed. 

```
>> version
ans =
    '9.4.0.813654 (R2018a)'
>> m = niftiinfo('sag_int_36sl_21.nii')
m = 
  struct with fields:
                 Filename: '/Users/rorden/dcm_qa/Ref/sag_int_36sl_21.nii'
              Filemoddate: '16-Jul-2018 16:55:03'
                 Filesize: 590176
              Description: 'TE=30;Time=135931.837;phase=1'
...
```

**Comment by Aghaeifar** [comment]
Thanks for your reply.
I already have MATLAB 9.4.0.813654 (R2018a) and tried to read [this nift](https://github.com/neurolabusc/dcm_qa/blob/master/Ref/ax_asc_35sl_6.nii)i from dcm_qa database 
Unfortunately again an empty array was returned by MATLAB . I tried on both windows and linux OS and got the same result.
'niftiinfo' works well with no problem. 
I also believe it should be something wrong from MATLAB  side, since all niftiviewers can open the files properly.

**Comment by neurolabusc** [comment]
The latest commit will support Matlab. However, the core problem is with Matlab, not dcm2niix. Specifically, the problem is with images.internal.nifti.niftiImage, as it does not follow the [specification](https://nifti.nimh.nih.gov/pub/dist/src/niftilib/nifti1.h). Specifically, the number of voxels must be determined by using the product of dim[1..dim[0]]. The specification gives examples  of this and gives no description regarding the values loaded in dim[1+dim[0]..7]. The current Matlab code computes the product of dim[1..7] regardless of dim[0]. This fails if the unused dims have any value other than 1. In the case of dcm2niix, it set unused dimensions to zero, whereas dcm2nii set these to 1. 

The bug is in the file images.internal.nifti.niftiImage, specifically the line `imageSize = prod(headerStruct.dim(2:8));`. This line should `imageSize = prod(headerStruct.dim(2:1+headerStruct.dim(1)));`. Alternatively, one might want to guard against software that does not correctly fill dim[0] you could add a conditional for the case where dim[0] is less than 1 or greater than 7:

```
 dim1 = headerStruct.dim(2:8);
 dim1(dim1(:) < 1) = 1;
 imageSize = prod(dim1);
```

**Comment by Aghaeifar** [comment]
Thanks, now it works fine.
I modified that MATLAB class according to your suggestion. 
Also this line needs to change:
`img = (reshape(img, headerStruct.dim(2:8)));`
to
`img = (reshape(img, headerStruct.dim(2:1+headerStruct.dim(1)))); `

**Comment by aandreev0** [comment]
Thank you for this discussion, saved me tons of time.
I made repo with this fix https://github.com/aandreev0/niftiImage-matlab/

It is private (due to MATLAB license) but I ma happy to share

**Comment by genericptr** [comment]
This bug was reported to Mathworks in July last year, and has been fixed in Matlab version 9.5.0.944444 (R2018b). So this issue only influences users running older versions of Matlab. Glad to here this bug report helped you.
